pkgpanda/http/__init__.py

```diff
@@ -0,0 +1,170 @@
+"""Pkgpanda HTTP API"""
+
+import http.client
+import logging
+import sys
+
+from flask import Flask, current_app, jsonify, request
+
+from pkgpanda import Install, Repository, actions
+from pkgpanda.exceptions import PackageError, ValidationError
+
+
+empty_response = ('', http.client.NO_CONTENT)
+
+
+def package_listing_response(package_ids):
+    return jsonify(sorted(package_ids))
+
+
+def package_response(package_id, repository):
+    try:
+        package = repository.load(package_id)
+    except PackageError:
+        error_message = 'Unable to load package {}.'.format(package_id)
+        logging.exception(error_message)
+        return error_response(error_message), http.client.INTERNAL_SERVER_ERROR
+
+    return jsonify({
+        'id': str(package.id),
+        'name': str(package.name),
+        'version': str(package.version),
+    })
+
+
+def error_response(message):
+    return jsonify({'error': message})
+
+
+app = Flask(__name__)
+app.config.from_object('pkgpanda.http.config')
+app.config.from_envvar('PKGPANDA_HTTP_CONFIG', silent=True)
+
+
+@app.errorhandler(Exception)
+def uncaught_exception_handler(exc):
+    error_message = 'An unexpected error has occurred.'
+    logging.error(error_message, exc_info=exc)
+    return error_response(error_message), http.client.INTERNAL_SERVER_ERROR
+
+
+@app.before_request
+def set_app_attrs_from_config():
+    current_app.install = Install(
+        current_app.config['DCOS_ROOT'],
+        current_app.config['DCOS_CONFIG_DIR'],
+        current_app.config['DCOS_ROOTED_SYSTEMD'],
+        manage_systemd=True,
+        block_systemd=False)
+    current_app.repository = Repository(
+        current_app.config['DCOS_REPO_DIR'])
+
+
+@app.route('/', methods=['GET'])
+def get_package_list():
+    return package_listing_response(current_app.repository.list())
+
+
+@app.route('/<package_id>', methods=['GET'])
+def get_package(package_id):
+    if not current_app.repository.has_package(package_id):
+        return (
+            error_response('Package {} not found.'.format(package_id)),
+            http.client.NOT_FOUND,
+        )
+    return package_response(package_id, current_app.repository)
+
+
+@app.route('/<package_id>', methods=['POST'])
+def fetch_package(package_id):
+    try:
+        repository_url = request.json['repository_url']
+    except Exception:
+        return (
+            error_response(
+                'Request body must be a json object with a `repository_url` '
+                'key.'
+            ),
+            http.client.BAD_REQUEST,
+        )
+
+    actions.fetch_package(
+        current_app.repository,
+        repository_url,
+        package_id,
+        current_app.config['WORK_DIR'])
+
+    return empty_response
+
+
+@app.route('/<package_id>', methods=['DELETE'])
+def remove_package(package_id):
+    if not current_app.repository.has_package(package_id):
```

Comment:
Succeeding doesn't sound like a bad thing in the remove non-existent case. The remove did succeed, in that the end state (package is not on host) has been reached. That's where the original CLI design came from.

In the context of users of the API, having remove be successful if it removes or the package already exists, makes it into a bit of a CRDT / completely repeatable in parallel without error operation. If two processes in parallel remove the same package, they'll both always succeed (rather than one getting a NOT_FOUND and having to decide that's just fine). Same goes for a co-ordination node.

Technically not found is slightly more information, although I'd rather just expose it as data inside of success (so people can operate on it if they want / be strict), but making it another error case just means people have to write more code to deal with it without a lot of value.